src/main/java/OurHashMap.java

```diff
@@ -1,75 +1,177 @@
+            }
+        }
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        for(List<Entry> list : values){
+            if(!(list == null)){
+                if(!list.isEmpty()){
+                    return false;
+                }
+            }
+        }
+        return true;
     }
 
     @Override
     public boolean containsKey(Object key) {
+        int hashcode = key.hashCode();
+        int index = Math.abs(hashcode) % SIZE;
+        List<Entry> list = values[index];
+        if(list == null){
+            return false;
+        }
+        for(Entry entry : list){
+            if(entry.key.equals(key)){
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public boolean containsValue(Object value) {
+        for(List<Entry> list : values){
+            if(!(list == null)){
+                for(Entry entry : list){
+                    if(entry.value.equals(value)){
+                        return true;
+                    }
+                }
+            }
+        }
         return false;
     }
 
     @Override
-    public Object get(Object key) {
+    public V get(Object key) {
         int hashcode = key.hashCode();
         int index = Math.abs(hashcode) % SIZE;
-        return values[index];
+        List<Entry> list = values[index];
+        if(list == null){
+            return null;
+        }
+        for(Entry entry : list){
+            if(entry.key.equals(key)){
+                return (V) entry.value;
+            }
+        }
+        return null;
     }
 
     @Override
-    public Object put(Object key, Object value) {
+    public V put(K key, V value) {
         int hashcode = key.hashCode();
         int index = Math.abs(hashcode) % SIZE;
-        values[index] = value;
-        // todo return previous value
+        List list = values[index];
+        if(list == null){
+            list = new ArrayList<Entry>();
+            values[index] = list;
+        }
+        for(Entry<K,V> entry : (List<Entry<K,V>>) list){
+            if(entry.key.equals(key)) {
+                V saved = entry.value;
+                entry.value = value;
+                return saved;
+            }
+        }
+
+        Entry entry = new Entry(key, value);
+        list.add(entry);
         return null;
     }
 
     @Override
-    public Object remove(Object o) {
+    public V remove(Object key) {
+        int hashcode = key.hashCode();
+        int index = Math.abs(hashcode) % SIZE;
+        List<Entry> list = values[index];
+        if(!(list == null)){
+            for(Entry entry : list){
+                if(entry.key.equals(key)){
+                    V value = (V) entry.value;
+                    list.remove(entry);
+                    return value;
+                }
+            }
+        }
         return null;
     }
 
     @Override
     public void putAll(Map map) {
-
+        Set<K> keys = map.keySet();
```

Comment:
Is there a way to do this more efficiently?

Reply:
I couldn't think of another way to access the entries in the map